projects/wizard_of_wikipedia/generator/tasks.py

```diff
@@ -0,0 +1,225 @@
+
+                    self.episodes.append({
+                        'chosen_topic': title,
+                        'title': title,
+                        'checked_sentence': best_ctx,
+                        'knowledge': (
+                            title + ' ' + TOKEN_KNOWLEDGE + ' ' + best_ctx
+                        ),
+                        'text': title + '\n' + question,
+                        'labels': list(ctx_ans[best_ctx]),
+                        'episode_done': True,
+                    })
+        if num_skipped:
+            import warnings
+            warnings.warn("Skipped {} squad questions in {}".format(num_skipped, path))
+
+    def get(self, episode_idx, entry_idx=None):
+        return self.episodes[episode_idx]
+
+
+class WizardTeacher(WizardDialogKnowledgeTeacher):
+    def __init__(self, opt, shared=None):
+        opt['label_type'] = 'response'
+        opt['include_checked_sentence'] = True
+        opt['task'] = 'foobar:' + opt['task']  # make sure the topic_split logic works
+        super().__init__(opt, shared)
+        self.knowledge_separator = True
+        self.only_checked_knowledge = opt.get('only_checked_knowledge', False)
+
+    @staticmethod
+    def add_cmdline_args(argparser):
+        WizardDialogKnowledgeTeacher.add_cmdline_args(argparser)
+        argparser.add_argument(
+            '--only-checked-knowledge', type='bool', default=False,
+            help='If true, only the checked sentence is provided'
+        )
+
+    def getID(self):
+        return "WizTeacher"
+
+    def get(self, episode_idx, entry_idx=0):
+        a = super().get(episode_idx, entry_idx)
+        # zero out the label candidates?
+        if 'knowledge' not in a:
+            # just a batch padding item
+            return a
+        # save some memory, we don't need label_candidates
+        a['label_candidates'] = []
+        if not a['knowledge'].startswith(TOKEN_NOCHOSEN):
+            # make sure the token is appearing
+            a['knowledge'] = (
+                TOKEN_NOCHOSEN + ' ' + TOKEN_KNOWLEDGE + ' ' + TOKEN_NOCHOSEN +
+                '\n' + a['knowledge']
+            )
+        if a['checked_sentence'] not in a['knowledge']:
+            # just default to no passage chosen
+            a['checked_sentence'] = TOKEN_NOCHOSEN
+            a['title'] = TOKEN_NOCHOSEN
+            #  but hard error if this isn't training time
+            if 'train' not in self.datatype:
+                raise ValueError(
+                    "Our checked_sentence exception *cannot* be allowed in test/valid."
+                )
+        if self.only_checked_knowledge:
+            # useful for test time evaluation, where it's only ever trained on true
+            # knowledge
+            a['knowledge'] = (
+                a['title'] + ' ' + TOKEN_KNOWLEDGE + ' ' + a['checked_sentence']
+            )
+        return a
+
+
+class GoldWizardTeacher(WizardTeacher):
```

Comment:
is it preferred in parlai to have a more very specific teachers, or one general teacher that can take in different parameters like 'only_checked_knowledge' or 'ignorant_dropout'?

Reply:
I’m not sure; CLI parameters are more common. This was just a convenient way for me to do it at the time.